### ISSUE TITLE
Dimensional check is incorrect in Compound.__sub__(), other possible latent problems

### ISSUE THREAD

**Issue by eecharlie** [gumyr/build123d#752]
```
from build123d import *
shape1 = Pos(0,0,0) * Box(30,30,30)
shape2 = Pos(30,0,0) * Box(50,5,5)
shape1 - shape2                                     # This is fine and produces a box with a hole
Compound(shape1) + Compound(shape2)                 # This is fine and produces a box on a stick
Compound([shape1,shape3]) + Compound(shape2)        # This works
sl = (ShapeList([shape1]) - ShapeList([shape2]))[0] # Executes, but just returns shape1, why?

Compound(children=[shape1]) - Compound(children=[shape2])   # This throws an error, why?
Compound(shape1) - Compound(shape2)                     # This throws an error, why?

# Error: TypeError: '<' not supported between instances of 'NoneType' and 'NoneType'
# when the dimension of self & other are checked
```

There are two problems quickly apparent in reviewing the source code for `__sub__()`:
```
        for _other in others:
            if type(_other)._dim < type(self)._dim:
                raise ValueError(
                    f"Only shapes with equal or greater dimension can be subtracted: "
                    ...
```
The error language is inconsistent with the actual conditional, in that the conditional should be `<=`, though this is not why the above example fails. The example fails because `type(Compound(Box(1,1,1)))._dim` is `None` when I think it should be `3`. 

The reason that this bug exists in `__sub__` and not in `__add__` is that add only checks for *equality* of dimension, and `None == None`. This is a significant separate latent issue. Here, fixing the conditional in `__sub__` in a careful way will allow it work, but at that point both `__add__` and `__sub__` and possibly other operations on `Compound` are silently failing to actually check dimensional compatibility of their arguments.

--> My plan is to fix the trivial bug in `__sub__()`'s dimension check such that it allows equality, knowing that this reproduces the latent failure in common with `__add__()`, but unblocks a use case for me. In a separate PR/issue, the fact that `Compound`'s dimension is `None` should be fixed.

**Comment by eecharlie** [comment]
Opened #754

**Comment by gumyr** [comment]
A Compound is a collection of any Shape; one can build a Compound of mixed 2D and 3D shapes so the `_dim` of Compound is explicitly set to None.  The `Part`, `Sketch` and `Curve` versions of `Compound` are intended to be used for collections of 3D, 2D and 1D Shapes respectively.

However, your other findings are worth investigation.

**Comment by eecharlie** [comment]
I think it’s a bit simpler than I first thought, see the PR.

**Comment by eecharlie** [comment]
So, why was I taking the approach in the above failing example in the first place? I think it is because the first and only place I recall seeing an explanation of the basic classes in the documentation is here:
https://build123d.readthedocs.io/en/latest/key_concepts.html#topology which includes `Compound` but none of `Part` `Sketch` or `Curve` classes. I think that an additional possibly confusing subtlety for me is the inclusion of `Shape` at the very end - I understand that it's the base class but I don't understand what I'd do with it. Should go at the beginning of the list of things that inherit from it? 

On the analogous page for Algebra mode, https://build123d.readthedocs.io/en/latest/key_concepts_algebra.html it *mentions* `Part` `Sketch` and `Curve` but there's no context on how one would obtain instances of them except implicitly from the examples that use function like `Box` and `Cylinder` which I don't think are documented anywhere outside of the weeds of the full API documentation.

Together, this documentation left me believing that if I wanted to manipulate a collection of disjoint 3D objects, I needed to use a `Compound`. (I'm still not positive that that's not the case - can a `Part` include physically disjoint components?)

I might suggest that the Topology section of Key Concepts (Builder Mode) be pulled out and precede both Key Concepts sections, because it's relevant to both. Further, I would suggest including in a single place a brief definition of `Part` `Sketch` `Curve` and explanation of their inheritance from `Compound`. It should be made clear that both `Compound` and `Shape` are not meant for direct use. Importantly, it should be clarified whether these children of `Compound` are meant to handle _collections_ of what the class name implies they are.
Additionally, a really helpful section complementing Topology and prior to Algebra/Builder mode would be a quick rundown of the helpful primitive `Part` `Sketch` and `Curve` creators, such as `Box` and `Cylinder`, even if it's just a laundry list of their names under the description of e.g. `Part` with links to the API documentation.

If there's some page in the documentation I missed that contains the above content, please point me to it, I've been unsuccessful with text searches for class names.

I would be open to taking on the above proposed documentation editing as another PR if you'd like. I barely know what I'm talking about but doing so would be a way to verify my knowledge :) if less efficient than someone intimately familiar with the codebase - however not being familiar with the codebase is my advantage here.

**Comment by gumyr** [comment]
`Compound` is definitely intended to be used, one of the ways is for an assembly - a collection of physically disjoint components - as described here: https://build123d.readthedocs.io/en/latest/assemblies.html  An `Assembly` class will be introduced as a subclass of `Compound` to help make this use case distinction.

`Part` can contain multiple disjoint objects as well but this isn't the primary purpose.  Often in the creation of an single `Solid` a intermediate stage may contain multiple `Solids` so this is not prohibited.

I've marked this Issue with the documentation tag, some of the recommendations you make here would likely be helpful for new users.

**Comment by eecharlie** [comment]
Ah, thanks. Will go back to understanding how my use case is better approached, informed by this, and see if I can understand your suggestion in the PR on minuend.

**Comment by gumyr** [comment]
`Compound.unwrap` just strips extraneous layers of `Compound` from an object.  With the `unwrap(fully=True)` option, the entire `Compound` "wrapper" is eliminated if there is a single Shape in the `Compound` (which OCCT generates frequently).  As both `self` and `others` are going to be combined and a new object generated these wrappers add no value.

**Comment by gumyr** [comment]
These operators handle Compounds now.
